File: core/src/main/java/no/nav/pensjon/vtp/felles/KeystoresGenerator.java
package no.nav.pensjon.vtp.felles;

import org.bouncycastle.asn1.x500.X500Name;
import org.bouncycastle.cert.X509CertificateHolder;
import org.bouncycastle.cert.X509v1CertificateBuilder;
import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
import org.bouncycastle.cert.jcajce.JcaX509v1CertificateBuilder;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.io.*;
import java.math.BigInteger;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.security.*;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.Date;

@Component
class KeystoresGenerator {
    private static final Logger log = LoggerFactory.getLogger(KeystoresGenerator.class);
    // TODO: Maybe use PKCS12? Java 9 and onwards uses that, Java 8 or below uses JKS.
    // String outputFormat = "PKCS12";

    private final CryptoConfigurationParameters cryptoConfigurationParameters;

    KeystoresGenerator(CryptoConfigurationParameters cryptoConfigurationParameters) {
        this.cryptoConfigurationParameters = cryptoConfigurationParameters;
    }

    void copyKeystoreAndTruststore() throws IOException {
        copyFile(cryptoConfigurationParameters.getKeystoreFilePath(), "keystore.jks");
        copyFile(cryptoConfigurationParameters.getTruststoreFilePath(), "truststore.jks");
    }

    private void copyFile(String filePath, String fileName) throws IOException {
        String serverDir = Paths.get("").toAbsolutePath().toString();

        Path source = Paths.get(serverDir, fileName);
        Path target = Paths.get(filePath);
        Files.createDirectories(target.getParent());
        Path path = Files.copy(source, target, StandardCopyOption.REPLACE_EXISTING);

        log.info(String.format("Copied %s/%s from project to %s (with replacing existing)", serverDir, fileName, path.toAbsolutePath()));
    }

    public void readKeystoresOrGenerateIfNotExists(String outputFormat, String keyAndCertAlias) {
        String keystorePath = cryptoConfigurationParameters.getKeystoreFilePath();
        String truststorePath = cryptoConfigurationParameters.getTruststoreFilePath();
        readKeystoresOrGenerateIfNotExists(keystorePath, cryptoConfigurationParameters.getKeyStorePassword(), truststorePath, cryptoConfigurationParameters.getTruststorePassword(), outputFormat, keyAndCertAlias);
    }

    public static void readKeystoresOrGenerateIfNotExists(String keystorePath, String keystorePassword, String truststorePath, String truststorePassword, String outputFormat, String keyAndCertAlias) {
        File keystoreFile = new File(keystorePath);
        keystoreFile.getParentFile().mkdirs();

        if (keystoreFile.length() == 0) {
            log.warn("Keystore keystoreFile {} does not exist - will auto-generate.", keystorePath);
            log.warn("OBS! Generated SAML signature from the new private key will be unknown/new for PEN/POPP and will not be accepted");
            createKeystoreAndUpdateTrustStore(keyAndCertAlias, keystorePath, keystorePassword, outputFormat, truststorePath, truststorePassword);
            return;
        }

        KeyStore.PrivateKeyEntry certEntry = null;
        try (FileInputStream keystoreFileIS = new FileInputStream(new File(keystorePath))) {
            KeyStore keyStore = KeyStore.getInstance(outputFormat);
            keyStore.load(keystoreFileIS, keystorePassword.toCharArray());
            KeyStore.ProtectionParameter protParam = new KeyStore.PasswordProtection(keystorePassword.toCharArray());
            certEntry = (KeyStore.PrivateKeyEntry) keyStore.getEntry(keyAndCertAlias, protParam);
        } catch (Exception e) {
            log.error("Could not read existing keystore. " + e.getMessage(), e);
        }
        if (certEntry == null) { //will not regenerate existing certificate
            createKeystoreAndUpdateTrustStore(keyAndCertAlias, keystorePath, keystorePassword, outputFormat, truststorePath, truststorePassword);
        }
    }

    private static void createKeystoreAndUpdateTrustStore(String keyAndCertAlias, String keystorePath, String keystorePassword, String outputFormat, String
            truststorePath, String truststorePassword) {
        log.info("Generating keystore {} with certificate for {}", keystorePath, keyAndCertAlias);
        Certificate selfCert = generateKeystore(keystorePath, keystorePassword, outputFormat, keyAndCertAlias);
        updateTruststore(keyAndCertAlias, outputFormat, truststorePath, truststorePassword, selfCert);
    }

    private static void updateTruststore(String keyAndCertAlias, String outputFormat, String truststorePath, String truststorePassword, Certificate selfCert) {
        try {
            if (!Files.exists(Paths.get(truststorePath))) {
                log.warn("Truststore file {} does not exist - will auto-generate.", truststorePath);
                File file = new File(truststorePath);
                file.getParentFile().mkdirs();
                file.createNewFile();
            }
            // Create the truststore, add the certificate, save to file
            log.info("Adding certificate to the truststore {}", truststorePath);
            KeyStore trustStore = KeyStore.getInstance(outputFormat);
            try (FileInputStream trustStoreFileIS = new FileInputStream(new File(truststorePath))) {
                InputStream stream = new File(truststorePath).length() > 0 ? trustStoreFileIS : null;
                trustStore.load(stream, truststorePassword.toCharArray());
            }
            trustStore.setCertificateEntry(keyAndCertAlias, selfCert);
            writeKeystoreToFile(trustStore, truststorePath, truststorePassword);
        } catch (Exception e) {
            log.error("Could not add generated certificate to trustStore: " + e.getMessage(), e);
            throw new RuntimeException(e);
        }
    }

    static Certificate generateKeystore(String keystorePath, String keystorePassword, String outputFormat, String keyAndCertAlias) {
        Security.addProvider(new BouncyCastleProvider());
        java.security.cert.Certificate selfCert;

        try {
            // Generate public/private keypair and self-signed certificate
            java.security.KeyPairGenerator keyPairGenerator = KeyPairGenerator
                    .getInstance("RSA");
            keyPairGenerator.initialize(2048);
            KeyPair keyPair = keyPairGenerator.generateKeyPair();
            PublicKey publicKey = keyPair.getPublic();
            PrivateKey privateKey = keyPair.getPrivate();
            selfCert = createMasterCert(
                    publicKey, privateKey);

            // Create the keystore, add the two entries
            // (we just use the same private key twice, for simplicity...),
            // save to file
            java.security.cert.Certificate[] outChain = {selfCert};
            KeyStore outStore = KeyStore.getInstance(outputFormat);
            outStore.load(null, keystorePassword.toCharArray());
            outStore.setKeyEntry(keyAndCertAlias, privateKey, keystorePassword.toCharArray(),
                    outChain);
            outStore.setKeyEntry("app-key", privateKey, keystorePassword.toCharArray(),
                    outChain);
            writeKeystoreToFile(outStore, keystorePath, keystorePassword);
        } catch (Exception e) {
            e.printStackTrace();
            throw new AssertionError(e.getMessage());
        }
        return selfCert;
    }

    private static void writeKeystoreToFile(KeyStore keyStore, String filePath, String password)
            throws IOException, CertificateException, NoSuchAlgorithmException, KeyStoreException {
        OutputStream outputStream = new FileOutputStream(filePath);
        keyStore.store(outputStream, password.toCharArray());
        outputStream.flush();
        outputStream.close();
    }

    private static Certificate createMasterCert(
            PublicKey pubKey,
            PrivateKey privKey)
            throws Exception {
        String issuer = "CN=vtp-pensjon";
        String subject = "CN=vtp-pensjon";

        // Valid from January 1st 1970 until ten years from now
        Date validFrom = new Date(0);
        Date validUntil = new Date(System.currentTimeMillis() + (1000L * 60 * 60 * 24 * 3650));

        X509v1CertificateBuilder builder = new JcaX509v1CertificateBuilder(
                new X500Name(issuer),
                BigInteger.valueOf(1),
                validFrom,
                validUntil,
                new X500Name(subject),
                pubKey
        );

        X509CertificateHolder holder = builder.build(
                new JcaContentSignerBuilder("SHA1WithRSA")
                        .setProvider("BC")
                        .build(privKey)
        );

        X509Certificate cert = new JcaX509CertificateConverter().setProvider("BC").getCertificate(holder);
        cert.checkValidity(new Date());
        cert.verify(pubKey);

        return cert;
    }
}

File: vtp-pensjon-application/src/test/java/no/nav/pensjon/vtp/application/KeystoresGenerator.kt
package no.nav.pensjon.vtp.application

import org.bouncycastle.asn1.x500.X500Name
import org.bouncycastle.jce.provider.BouncyCastleProvider
import org.bouncycastle.cert.X509v1CertificateBuilder
import org.bouncycastle.cert.jcajce.JcaX509v1CertificateBuilder
import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder
import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter
import java.io.*
import java.math.BigInteger.valueOf
import java.nio.file.Files
import java.security.*
import java.security.KeyStore.getInstance
import java.security.Security.addProvider
import java.security.cert.Certificate
import java.util.*

/**
 * Dette brukes bare for å generere et nytt nøkkel-par.
 * 1. Generer nøkkel-par ved å kjøre main-metoden
 * 2. Eksporter sertifikat etter generering:
 * keytool -export -keystore ~/.modig/keystore.jks -alias localhost-ssl -file ~/.modig/ vtp-pensjon.cer
 * 3. Importer sertifikaten inn i PEN lokale truststores som ligger i pesys/local/pen-secrets:
 * keytool -import -alias localhost -file ~/.modig/vtp-pensjon.cer -storetype JKS -keystore ~/pesys/local/pen-secrets/truststore/truststore.jts
 * 4. Importer sertifikaten inn i POPP lokale truststores som ligger i pesys/local/popp-secrets:
 * keytool -import -alias localhost -file ~/.modig/vtp-pensjon.cer -storetype JKS -keystore ~/pesys/local/popp-secrets/truststore/truststore.jts
 * 5. Lag en Pull-Request til pesys-master med oppdaterte lokale truststores
 * 6. Kopier og erstatt vtp truststore og keystore i keystore.jks og truststore med nye genererte
 * 7. Lag Pull-Request til vtp-master med oppdaterte truststore/keystore
 */
fun main(args: Array<String>) {
    val keystoreFile = File(System.getProperty("user.home", ".") + "/.modig/keystore.jks")
    val keystorePassword = "devillokeystore1234"
    val truststoreFile = File(System.getProperty("user.home", ".") + "/.modig/truststore.jks")
    val truststorePassword = "changeit"
    val keyAndCertAlias = "localhost-ssl"

    println("----------- GENERATING A NEW KEY PAIR -----------")
    Files.deleteIfExists(keystoreFile.toPath())
    Files.deleteIfExists(truststoreFile.toPath())
    keystoreFile.parentFile.mkdirs()
    truststoreFile.parentFile.mkdirs()

    createKeystoreAndTrustStore(keyAndCertAlias, keystoreFile, keystorePassword.toCharArray(), "JKS", truststoreFile, truststorePassword.toCharArray())
    println("-------------------- DONE -----------------------")
}

fun createKeystoreAndTrustStore(keyAndCertAlias: String, keystoreFile: File, keystorePassword: CharArray, outputFormat: String, truststorePath: File, truststorePassword: CharArray) {
    println("Generating keystore $keystoreFile with certificate for $keyAndCertAlias")
    val selfCert = generateKeystore(keystoreFile, keystorePassword, outputFormat, keyAndCertAlias)
    createTruststore(keyAndCertAlias, outputFormat, truststorePath, truststorePassword, selfCert)
}

private fun createTruststore(keyAndCertAlias: String, outputFormat: String, truststoreFile: File, truststorePassword: CharArray, selfCert: Certificate) {
    // Create the truststore, add the certificate, save to file
    println("Adding certificate to the truststore $truststoreFile")
    getInstance(outputFormat).apply {
        load(null, truststorePassword)
        setCertificateEntry(keyAndCertAlias, selfCert)
        writeKeystoreToFile(this, truststoreFile, truststorePassword)
    }
}

fun generateKeystore(keystoreFile: File, keystorePassword: CharArray, outputFormat: String?, keyAndCertAlias: String?): Certificate {
    addProvider(BouncyCastleProvider())

    // Generate public/private keypair and self-signed certificate
    val keyPair = KeyPairGenerator
            .getInstance("RSA").apply {
                initialize(2048)
            }.generateKeyPair()
    val selfCert = createMasterCert(keyPair.public, keyPair.private)

    // Create the keystore, add the two entries
    // (we just use the same private key twice, for simplicity...),
    // save to file
    val outChain = arrayOf(selfCert)
    getInstance(outputFormat).apply {
        load(null, keystorePassword)
        setKeyEntry(keyAndCertAlias, keyPair.private, keystorePassword, outChain)
        setKeyEntry("app-key", keyPair.private, keystorePassword, outChain)
        writeKeystoreToFile(this, keystoreFile, keystorePassword)
    }
    return selfCert
}

private fun writeKeystoreToFile(keyStore: KeyStore, file: File, password: CharArray) {
    file.outputStream().use { keyStore.store(it, password) }
}

private fun createMasterCert(pubKey: PublicKey, privKey: PrivateKey): Certificate {
    val issuer = "CN=vtp-pensjon"
    val subject = "CN=vtp-pensjon"

    // Valid from January 1st 1970 until ten years from now
    val validFrom = Date(0)
    val validUntil = Date(System.currentTimeMillis() + 1000L * 60 * 60 * 24 * 3650)

    val builder: X509v1CertificateBuilder = JcaX509v1CertificateBuilder(
            X500Name(issuer),
            valueOf(1),
            validFrom,
            validUntil,
            X500Name(subject),
            pubKey
    )

    return JcaX509CertificateConverter()
            .setProvider("BC")
            .getCertificate(builder.build(
                    JcaContentSignerBuilder("SHA1WithRSA")
                            .setProvider("BC")
                            .build(privKey)
            ))
            .apply {
                checkValidity(Date())
                verify(pubKey)
            }
}
